tidy3d/web/core/environment.py

```diff
@@ -1,10 +1,9 @@
 """Environment Setup."""
 import os
+import logging as log
```

Comment:
@tylerflex, I forgot, is `web/core` not supposed to import tidy3d components?

Reply:
hmm, this is actually related to what we were discussing above about using `get_logging_console()`. I guess if we are adamant on not importing any tidy3d functionality in `web/core` then we have to somehow make sure that formatting (currently just console width) is consistent between tidy3d and `web/core`. Or, alternatively, we can relax this restriction and allow import of the top level tidy3d components, e.g., `tidy3d.log`, `get_logging_console()`, etc

Reply:
I think I missed this discussion, so I guess my comment is incorrect in that case. However, I think top-level modules should be accessible to `web.core`, so that exceptions and logging are shared and consistent across the whole module.

Reply:
the problem is that we were considering making `web/core` its own separate package potentially. However, I agree that if we keep it as part of `tidy3d`, then we could give it access to the top level stuff like exceptions, logging (even `__version__`). Having it as a separate package makes development easier between the EM and MC teams though as we can work independently and then just sync up after releases. Do you have any preferences or thoughts regarding the decision between keeping `web/core` in tidy3d vs having it separate?

Reply:
well, even as a separate repo, it could still import tidy3d, right? Like denormalizer, for example. Although there it is a very crucial component, not just for logging

Reply:
If it is possible to import tidy3d, will it create circular dependencies? If it is split into 2 repositories, how will this dependency be decoupled?

Reply:
Python (pip, conda, etc.) handles cyclic dependencies between packages just fine as far I know. It wouldn't be a problem to have tidy3d and web being separate, co-dependent packages. OTOH, that's not the most elegant solution, which would be to have a 3rd base package with the core functionality (log, exceptions, etc.).

I'm not 100% aware of what would be the duplicated functionality between web and tidy3d right now, but if it is only logging, we could have the core module accept a `log` object that would be used internally and that we can set when importing it from tidy3d. For example:

```python
# In tidy3d.__init__.py
import tidy3d_core
from .log import log

tidy3d_core.set_logger(log)
```

Reply:
I think I'm in favor of ensuring that `web/core` does not import from tidy3d at all just in case we want to make it a separate package eventually without worrying about circular dependency. So perhaps the best approach to logging is to handle it how Lucas suggests here and provide an API for `web/core` to set its logger.  Maybe we can do something similar for the console (`tidy3d_core.set_console_getter(get_console)`)

Reply:
Sounds good to me. Any arguments for setting a console getter vs a console instance `.set_console(get_logging_console())`? also, should log and console have some default values if `.set_logger()` and `.set_console()` are not called?

Reply:
I suggested this because we call `get_logging_console()` in the webapi instead of importing a `Console` instance. I think it needs to be this way because the `console` represents a specific output (rather than `log` which is more a handler for output). But I'm not actually sure on this, if we can get away with a specific console instance, that might be easier than setting a getter function even in our webapi.

Reply:
tidy3d
```python
# In tidy3d.__init__.py
from .core import core_log
from ..log import log, get_logging_console

core_log.set_logger(log, get_logging_console())
```

tidy3d/core add core_log.py

```python
logger_setting = {
    "logger": "",
    "logger_console": "",
}

def set_logger(logger, logger_console):
    logger_setting["logger"] = logger
    logger_setting["logger_console"] = logger_console
```

So, s3util can use 
```python

def _get_progress(action: _S3Action):
    """Get the progress of an action."""

    col = (
        TextColumn(f"[bold green]{_S3Action.DOWNLOADING.value}")
        if action == _S3Action.DOWNLOADING
        else TextColumn(f"[bold red]{_S3Action.UPLOADING.value}")
    )
    return Progress(
        col,
        TextColumn("[bold blue]{task.fields[filename]}"),
        BarColumn(),
        "[progress.percentage]{task.percentage:>3.1f}%",
        "•",
        DownloadColumn(),
        "•",
        TransferSpeedColumn(),
        "•",
        TimeRemainingColumn(),
        console=get_logger_console(),
    )
```
Merged the code to stub, have a check, any discuss pls let me know.

Reply:
I think this looks pretty good. Only remaining question is whether it Is ok that the logging console is an instance here instead of a getter function. Can we test again running the WebAPI.ipynb notebook to ensure that it formats correctly? if so I think I'm happy with this change.

Small note: I think `get_logger_console()` will already return `None` if no logger is set, so maybe we don't need the `or None` part in `console=get_logger_console() or None`?

Reply:
Not sure if the notebook will work until the corresponding changes on web side will be done. But yeah, passing a getter function is probably safer. In fact, `set_logging_console()` method from `tidy3d.log` changes console instance returned by `get_logging_console()`.

Reply:
yea, changed `console=get_logger_console() or None` to `console=get_logger_console()`